doc/stm/index.mld

```diff
@@ -0,0 +1,581 @@
+[next_state cmd state]). This is useful because we can now check that the result of
+the [remove] function is correct according to the model's [state].
+
+{[
+  let postcond cmd state res =
+    match cmd, res with
+    | Mem i,  Res ((Bool,_), b)               -> b = List.mem i state
+    | Cardinal, Res ((Int,_), l)              -> l = List.length state
+    | Add _,  Res ((Unit,_),_)                -> true
+    | Remove i, Res ((Option Int, _), Some x) -> List.mem i state && i = x
+    | Remove i, Res ((Option Int, _), None)   -> not (List.mem i state)
+    | _                                       -> false
+]}
+
+Here, we can pattern match on the result of the [Remove i] [cmd] and express
+the fact that if this result is [Some x] then [i] was indeed in the set and [i]
+and [x] are structurally equal. On the other hand, if the result is [None], then
+[i] was not in the set.
+
+Now, let's say we have updated [arb_cmd] just by adding a generator for the [Remove]
+[cmd] similar to the other ones:
+
+{[
+  let arb_cmd _state =
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Cardinal;
+         Gen.map (fun i -> Mem i) Gen.int;
+         Gen.map (fun i -> Add i) Gen.int;
+         Gen.map (fun i -> Remove i) Gen.int;
+        ])
+]}
+
+Then, if we make the same mistake as above for the [add] function, namely:
+
+{[
+  let remove a t =
+    Mutex.lock t.mutex;
+    let r =
+      if mem_non_lock a t then begin
+        t.content <- S.remove a t.content;
+        (* t.cardinal <- t.cardinal - 1; *)
+        Some a
+      end
+      else None
+    in
+    Mutex.unlock t.mutex;
+    r
+]}
+
+most of the time, [qcheck-stm] won't be able to spot the bug:
+
+{[
+$ dune exec ./stm-doc/mutable_set_v4.exe
+random seed: 423411827
+generated error fail pass / total     time test name
+[✓]  100    0    0  100 /  100     0.8s STM sequential tests
+================================================================================
+success (ran 1 tests)
+]}
+
+The reason why is that there is very little chance to generate an argument for
+[Remove] corresponding to an element already added to the set in the same
+program. So the conditional branch where the mistake has been made is never
+explored. But be sure that it will be in real life!
+
+In order to overcome this problem, it is possible to skew the distribution of
+generated [cmd]s by looking at the [state] parameter. If the model of the set
+is empty, then we just generate a random element. But if there are already some
+elements in the set, then we can choose between picking one of them or generating a
+random new one:
+
+{[
+  let arb_cmd state =
+    let gen =
+      match state with
+      | [] -> Gen.int
+      | xs -> Gen.(oneof [oneofl xs; int])
+    in
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Cardinal;
+         Gen.map (fun i -> Mem i) Gen.int;
+         Gen.map (fun i -> Add i) Gen.int;
+         Gen.map (fun i -> Remove i) gen;
```

Comment:
By the same argument
- `Mem` will rarely ask for an actual member
- `Add` will rarely test for adding an existing member

For this reason I think they should also be updated to use the `gen` generator

Reply:
Yes, I agree.
And it is the occasion to add a small remark to draw the reader's attention on the fact the problem is not always obvious :-)

Reply:
I know of two approaches to help notice it:
- *fault injection*, e.g., if we alter `mem` to always return `false` we should wonder why tests still succeed with that change
- printing samples from the `cmd` generators distribution with `Gen.generate ~n:10` or `Gen.generate1`

It may be good to document these tricks at some point - but I think it is fair enough to leave them out for this first version.